sys/kern/src/arch/riscv32.rs

```diff
@@ -435,11 +435,17 @@ fn timer_handler() {
                 set_current_task(next);
             }
 
-            // Reset mtime back to 0.  In theory we could save an instruction on
-            // RV32 here and only write the low-order bits, assuming that it has
-            // been less than 12 seconds or so since our last interrupt(!), but
-            // let's avoid any possibility of a nasty surprise.
-            core::ptr::write_volatile(crate::startup::MTIME as *mut u64, 0);
+            //
+            // Increase mtimecmp for the next interrupt
+            // This will also clear the pending timer interrupt.
+            //
+            let mtimecmp =
+                core::ptr::read_volatile(crate::startup::MTIME as *mut u64);
+            let new_mtimecmp = mtimecmp + CLOCK_FREQ_KHZ as u64;
```

Comment:
This should probably be an [overflowing_add](https://doc.rust-lang.org/stable/core/primitive.u64.html#method.overflowing_add), otherwise Rust will put a panic check.

Reply:
Actually I'm not sure if these systems ever will run long enough for that to happen... might not be necessary. Either way, I'm pretty sure an overflow is the expected behavior so we might want to allow it.

Reply:
That is an interesting point. Yes, if the system runs long enough we can have a case where MTIME would overflow. I don't think we can make the assumption that the system won't run long enough. In those cases, wouldn't it be wrong to configure mtimecmp with the result of `overflowing_add` or just a simple add? Shouldn't we set it to u64::MAX? Else, the timer would keep firing interrupts until MTIME also wraps around?

Reply:
With other timer stuff, its was decided that letting the overflow panic is okay.  See the comment [here](https://github.com/rivosinc/hubris/blob/rivos/main/sys/kern/src/arch/riscv32.rs#L409).  

I think if we are concerned about the overflow, we would just not change to handling the timer this way.

What does everyone else think?

Reply:
Good point. Even if we consider a fast tick happening every ns, then it will take > 500 years to overflow. Should probably be fine to assume no system is going to be up for that long :-P. Having an `overflowing_add` might still be okay in that case though?

Reply:
Should we change how `TICKS` is handled then?  It is also setup such that it will overflow eventually.
If we set `MTIMECMP` to `u64::max` we would still have a timer glitch, where we would get interrupted on a shortened interval.

Reply:
> Should we change how TICKS is handled then? It is also setup such that it will overflow eventually.

No, the reasoning you provided seems fine to me w.r.t. not requiring special handling for the overflow. I updated my earlier comment to scratch out the suggestion.

> If we set MTIMECMP to u64::max we would still have a timer glitch, where we would get interrupted on a shortened interval.

Yes, there would be a timer glitch but it wouldn't lead to interrupts on every tick. Consider the case where lets say the overflow value is 32 and MTIME is at 25 and tick_divisor is at 8. Now MTIMECMP would be set to 1. This would cause interrupt to be fired at MTIME 26, 27, 28, 29, 30 and 31. Hence, I suggested we could use u64::MAX and have the timer fired only once. Yes, the interrupt will arrive sooner but if we don't have to worry about an early interrupt once in a long while, we might be okay. Anyways, I scratched out that suggestion, I don't think we need to worry about that for simplification.